### DISCUSSION TITLE
calt substitution problem

### DISCUSSION THREAD

**Discussion by rajeeshknambiar** [harfbuzz/harfbuzz#4050] [Q&A] [answered]
I’m not sure if this qualifies as a bug report yet; so discussing here first.

I have a contextual substitution which replaces `r3cil` glyph with `dotreph` in the attached font (RIT-Rachana):
```fea
lookup calt_dot_reph {
    sub r3cil' [... l3l3 ...] by dotreph;  #full list of context glyphs can be found in otf
} calt_dot_reph;
```

`hb-shape` indeed seems to substitute it correctly
```bash
hb-shape --font-file RIT-Rachana-Regular.otf "ര്‍ല്ല"   #r3  xx zwj l3 xx l3 → r3cil l3l3 → dotreph l3l3
[dotreph=0+0|l3l3=3+1450]
```
but `hb-view` is missing the `dotreph` glyph, only `l3l3` glyph is displayed.
```bash
hb-view --font-file RIT-Rachana-Regular.otf "ര്‍ല്ല" -o /tmp/hb-test.png   #expects dot reph above l3l3
```
Actual result: 
------------------
![image](https://user-images.githubusercontent.com/424259/213677939-09b7d42d-e8a8-469a-96b9-dfec74183af6.png)

Expected  result:
------------------
![image](https://user-images.githubusercontent.com/424259/213678038-11b2f3de-f666-46bb-be5e-f7d7f1591d4b.png)

Font: 
[RIT-Rachana-Regular.zip](https://github.com/harfbuzz/harfbuzz/files/10465584/RIT-Rachana-Regular.zip)

What am I missing?

**Comment by khaledhosny** [comment]
The glyphs is there but too far to the left, and hb-view bounds calculation misses this. Try:
```
hb-view --font-file RIT-Rachana-Regular.otf "ര്‍ല്ല"  --margin=10,10,10,200
```
![a](https://user-images.githubusercontent.com/93914/213694140-a6bd36b3-fd28-47fb-a900-a6cc656d0aaf.png)

If mark attachment should happen here, for some reason it isn’t working.

**Reply by rajeeshknambiar** [reply]
Ok, so the glyph is present. But `GPOS` doesn't seem to work, even if mark positioning is defined in the font (`abvm`).  `hb-shape -V` shows
```
...
trace: end table GSUB   buffer: [dotreph=0|l3l3=3]
trace: start table GPOS buffer: [dotreph=0+0|l3l3=3+1450]
trace: start lookup 0   buffer: [dotreph=0+0|l3l3=3+1450]
trace: end lookup 0     buffer: [dotreph=0+0|l3l3=3+1450]
trace: start lookup 1   buffer: [dotreph=0+0|l3l3=3+1450]
trace: end lookup 1     buffer: [dotreph=0+0|l3l3=3+1450]
trace: start lookup 2   buffer: [dotreph=0+0|l3l3=3+1450]
trace: end lookup 2     buffer: [dotreph=0+0|l3l3=3+1450]
trace: start lookup 3   buffer: [dotreph=0+0|l3l3=3+1450]
trace: end lookup 3     buffer: [dotreph=0+0|l3l3=3+1450]
trace: start lookup 4   buffer: [dotreph=0+0|l3l3=3+1450]
trace: end lookup 4     buffer: [dotreph=0+0|l3l3=3+1450]
trace: end table GPOS   buffer: [dotreph=0+0|l3l3=3+1450]
[dotreph=0+0|l3l3=3+1450]
```

**Comment by dscorbett** [comment, marked as answer]
It isn’t working because `dotreph` is not being reordered towards the end of the syllable. That is because HarfBuzz doesn’t recognize it as a reph. The reph should be encoded as U+0D4E MALAYALAM LETTER DOT REPH, not as <U+0D30, U+0D4D, U+200D>. [<i>The Unicode Standard</i>, version 15.0, chapter 12](https://www.unicode.org/versions/Unicode15.0.0/ch12.pdf#page=69) says “The sequence <<i>ra</i>, <i>virama</i>, ZWJ> is not used to represent the <i>dot reph</i>, because that sequence has considerable preexisting usage to represent the <i>chillu</i> form of <i>ra</i>, prior to the encoding of the <i>chillu</i> form as a distinct character, U+0D7C U+0D7C MALAYALAM LETTER CHILLU RR.”

If the text follows Unicode’s recommendation, this font works:
```console
$ hb-shape RIT-Rachana-Regular.otf -u '0D4E 0D32 0D4D 0D32'
[l3l3=0+1450|dotreph=0@196,0+0]
```
![ൎല്ല](https://user-images.githubusercontent.com/1124347/213720475-7cf29c6d-50fc-4a7a-b06f-5199c695af7c.png)

**Reply by rajeeshknambiar** [reply]
The shaping is (already) correct with dotreph `0D4E` + conjuncts.
 
Dot reph is linguistically only a stylistic alternate of chillu-ra; when usually preceding double consonants (and a few specific conjuncts). I am trying to bring that behaviour in the font for such specific combinations using `calt` so that typesetting applications can opt into it.

As Khaled explained, the glyphs are correct, but their `GPOS` rules doesn't seem  to trigger.

**Reply by behdad** [reply]
> As Khaled explained, the glyphs are correct, but their `GPOS` rules doesn't seem to trigger.

Because the order of glyphs is different.

**Reply by rajeeshknambiar** [reply]
> > As Khaled explained, the glyphs are correct, but their `GPOS` rules doesn't seem to trigger.
> 
> Because the order of glyphs is different.

Could you elaborate? Both chillu-ra and dotreph precedes the base glyph; and using dotreph directly produces expected o/p:
```
hb-view --font-file RIT-Rachana-Regular.otf "ൎല്ല"   # dotreph l3 xx l3
```

**Reply by behdad** [reply]
The shaper moves dotreph to after the baseglyph, but not the chillu-ra.

**Reply by rajeeshknambiar** [reply]
> The shaper moves dotreph to after the baseglyph, but not the chillu-ra.

Ok, that explains.

Is there an advisable method to display a conjunct with dotreph positioned properly using `calt` for chillu-ra?

**Reply by behdad** [reply]
> Is there an advisable method to display a conjunct with dotreph positioned properly using `calt` for chillu-ra?

Since the glyph will be before the base, you cannot use mark-positioning. Your best bet is to draw the glyph in a way that it just is positioned in roughly the right place I'm afraid...

**Reply by rajeeshknambiar** [reply]
> The shaper moves dotreph to after the baseglyph, but not the chillu-ra.

Indeed, this happens at `reordering indic final step`:
```
trace: end lookup 10    buffer: [dotreph=0|l3l3=1]
trace: start reordering indic final     buffer: [dotreph=0|l3l3=1]
trace: end reordering indic final       buffer: [l3l3=0|dotreph=0]             ← reordered  here
trace: start lookup 11  buffer: [l3l3=0|dotreph=0]
trace: end lookup 11    buffer: [l3l3=0|dotreph=0]
trace: end table GSUB   buffer: [l3l3=0|dotreph=0]
trace: start table GPOS buffer: [l3l3=0+1450|dotreph=0+0]
trace: start lookup 0   buffer: [l3l3=0+1450|dotreph=0+0]
trace: end lookup 0     buffer: [l3l3=0+1450|dotreph=0+0]
trace: start lookup 1   buffer: [l3l3=0+1450|dotreph=0+0]
trace: attaching mark glyph at 1 to glyph at 0  buffer: [l3l3=0+1450|dotreph=0+0] ← mark attachment
trace: attached mark glyph at 1 to glyph at 0   buffer: [l3l3=0+1450|dotreph=0@1646,0+0]
trace: end lookup 1     buffer: [l3l3=0+1450|dotreph=0@1646,0+0]
trace: start lookup 2   buffer: [l3l3=0+1450|dotreph=0@1646,0+0]
trace: end lookup 2     buffer: [l3l3=0+1450|dotreph=0@1646,0+0]
trace: start lookup 3   buffer: [l3l3=0+1450|dotreph=0@1646,0+0]
trace: kerning glyphs at 0,1    buffer: [l3l3=0+1450|dotreph=0@1646,0+0]       ← kerning
trace: kerned glyphs at 0,1     buffer: [l3l3=0+1450|dotreph=0@1646,0+0]
trace: end lookup 3     buffer: [l3l3=0+1450|dotreph=0@1646,0+0]
trace: start lookup 4   buffer: [l3l3=0+1450|dotreph=0@1646,0+0]
trace: end lookup 4     buffer: [l3l3=0+1450|dotreph=0@1646,0+0]
trace: end table GPOS   buffer: [l3l3=0+1450|dotreph=0@1646,0+0]
```

**Reply by rajeeshknambiar** [reply]
> > Is there an advisable method to display a conjunct with dotreph positioned properly using `calt` for chillu-ra?
> 
> Since the glyph will be before the base, you cannot use mark-positioning. Your best bet is to draw the glyph in a way that it just is positioned in roughly the right place I'm afraid...

It *looks* like reordering could work with chaining substitution (ref. *Example 5* in Section *5.f.i* of https://adobe-type-tools.github.io/afdko/OpenTypeFeatureFileSpecification.html#5.f.i). I presume `ka.pas_cakra.ns` in that example is removed/substituted away.

I could get the reordering work (change `r3cil l3l3` to `l3l3 dotreph`), but the final `l3l3` is not removed. Not sure whether it is an issue with Fontforge (which is what merges the feature file). These are the lookups I am trying to apply:
```

lookup REMOVE_l3l3 {
    sub r3cil l3l3 by r3cil;
} REMOVE_l3l3;

lookup REORDER_reph_l3l3 {
    sub r3cil by l3l3 dotreph;
} REORDER_reph_l3l3;

lookup calt_dot_reph {
    sub r3cil' lookup REMOVE_l3l3 lookup REORDER_reph_l3l3 l3l3' ;
} calt_dot_reph;

feature calt {
    lookup calt_dot_reph;
} calt;
```

And the shaping becomes:
```bash
[l3l3=0+1450|dotreph=0@196,0+0|l3l3=3+1450]    # l3l3 at the end is not substituted away
```

May be @simoncozens have some ideas?

**Comment by simoncozens** [comment]
Try this (untested)

```fea
lookup KILL_l3l3 {
   sub l3l3 by NULL;
} KILL_l3l3;

    sub r3cil' lookup REMOVE_l3l3 lookup REORDER_reph_l3l3 l3l3' lookup KILL_l3l3 ;
```

**Reply by rajeeshknambiar** [reply]
That still doesn't kill the last `l3l3`:
```bash
[l3l3=0+1450|dotreph=0@196,0+0|l3l3=3+1450]
```

The font is attached for reference: 
[RIT-Rachana-Regular.zip](https://github.com/harfbuzz/harfbuzz/files/10498128/RIT-Rachana-Regular.zip)

**Reply by simoncozens** [reply]
Wait, we're all making this too difficult. Looks like what you want to do is substitute the r3cil by l3l3 and l3l3 for dotreph. So surely the code is

```fea
lookup swap {
    sub r3cil by l3l3;
    sub l3l3l by dotreph;
} swap;

lookup calt_dot_reph {
    sub r3cil' lookup swap l3l3' lookup swap ;
} calt_dot_reph;
```

**Reply by rajeeshknambiar** [reply]
And... that is just perfect (barring a minor typo `l3l3l`).
```
...
trace: end lookup 10    buffer: [r3cil=0|l3l3=3]
trace: start reordering indic final     buffer: [r3cil=0|l3l3=3]
trace: end reordering indic final       buffer: [r3cil=0|l3l3=3]
trace: start lookup 12  buffer: [r3cil=0|l3l3=3]
trace: recursing to lookup 11 at 0      buffer: [r3cil=0|l3l3=3]
trace: replacing glyph at 0 (single substitution)       buffer: [r3cil=0|l3l3=3]  ← (sequence now: r3cil l3l3)
trace: replaced glyph at 0 (single substitution)        buffer: [l3l3=0|l3l3=3]   ← r3cil replaced (sequence now: l3l3 l3l3)
trace: recursed to lookup 11    buffer: [l3l3=0|l3l3=3]
trace: recursing to lookup 11 at 1      buffer: [l3l3=0|l3l3=3]
trace: replacing glyph at 1 (single substitution)       buffer: [l3l3=0|l3l3=3]
trace: replaced glyph at 1 (single substitution)        buffer: [l3l3=0|dotreph=3] ← second l3l3 replaced (sequence now: l3l3 dotreph)
trace: recursed to lookup 11    buffer: [l3l3=0|dotreph=3]
trace: end lookup 12    buffer: [l3l3=0|dotreph=3]
trace: end table GSUB   buffer: [l3l3=0|dotreph=3]
trace: start table GPOS buffer: [l3l3=0+1450|dotreph=3+0]
trace: start lookup 0   buffer: [l3l3=0+1450|dotreph=3+0]
trace: end lookup 0     buffer: [l3l3=0+1450|dotreph=3+0]
trace: start lookup 1   buffer: [l3l3=0+1450|dotreph=3+0]
trace: attaching mark glyph at 1 to glyph at 0  buffer: [l3l3=0+1450|dotreph=3+0]
trace: attached mark glyph at 1 to glyph at 0   buffer: [l3l3=0+1450|dotreph=3@1646,0+0]
trace: end lookup 1     buffer: [l3l3=0+1450|dotreph=3@1646,0+0]
trace: start lookup 2   buffer: [l3l3=0+1450|dotreph=3@1646,0+0]
trace: end lookup 2     buffer: [l3l3=0+1450|dotreph=3@1646,0+0]
trace: start lookup 3   buffer: [l3l3=0+1450|dotreph=3@1646,0+0]
trace: kerning glyphs at 0,1    buffer: [l3l3=0+1450|dotreph=3@1646,0+0]
trace: kerned glyphs at 0,1     buffer: [l3l3=0+1450|dotreph=3@1646,0+0]
trace: end lookup 3     buffer: [l3l3=0+1450|dotreph=3@1646,0+0]
trace: start lookup 4   buffer: [l3l3=0+1450|dotreph=3@1646,0+0]
trace: end lookup 4     buffer: [l3l3=0+1450|dotreph=3@1646,0+0]
trace: end table GPOS   buffer: [l3l3=0+1450|dotreph=3@1646,0+0]
[l3l3=0+1450|dotreph=3@196,0+0]
```

For a better understanding: does the lookup gets applied as follows?

1. `r3cil' lookup swap`  → triggers lookup and replaces `r3cil` with `l3l3`.
1. `l3l3' lookup swap`  → triggers lookup and replaces `l3l3` with `dotreph`.
Resulting in sequence `l3l3 dotreph` and GPOS processing takes care of  the rest.

**Reply by simoncozens** [reply]
Exactly!

**Reply by rajeeshknambiar** [reply]
Many thanks, Simon.

One (possibly final question):  similar substitution should take place for a number of conjuncts, the only option to make them all work is something as follows?
```fea
...
lookup swap_reph_y1y1 {
	sub r3cil by y1y1;
} swap_reph_y1y1;

lookup swap_reph_l3l3 {
	sub r3cil by l3l3;
} swap_reph_l3l3;

lookup swap_reph_v1v1 {
	sub r3cil by v1v1;
} swap_reph_v1v1;

lookup swap_reph {
    sub [k1k1 k3k3 ch1ch1 ch3ch3 nhnh th1th1 th3th3 th3th4 n1n1 p1p1 p3p3 m1m1 y1y1 l3l3 v1v1] by dotreph;
} swap_reph;

lookup calt_dot_reph {
    # swap r3cil with following conjunct, swap the conjunct with dotreph
    ...
    sub r3cil' lookup swap_reph_y1y1 y1y1'    lookup swap_reph ;
    sub r3cil' lookup swap_reph_l3l3 l3l3'    lookup swap_reph ;
    sub r3cil' lookup swap_reph_v1v1 v1v1'    lookup swap_reph ;
```

**Reply by simoncozens** [reply]
Sadly, yes.
